src/components/vm/consoles/consoles.jsx

```diff
@@ -17,143 +17,236 @@
  * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
  */
 import React from 'react';
-import PropTypes from 'prop-types';
 import cockpit from 'cockpit';
-import { AccessConsoles } from "@patternfly/react-console";
+import { StateObject } from './state';
+import { Button } from "@patternfly/react-core/dist/esm/components/Button";
+import { Card, CardBody, CardHeader, CardTitle } from '@patternfly/react-core/dist/esm/components/Card';
+import { ExpandIcon, CompressIcon } from "@patternfly/react-icons";
+import { ToggleGroup, ToggleGroupItem } from '@patternfly/react-core/dist/esm/components/ToggleGroup';
+import { Split, SplitItem } from "@patternfly/react-core/dist/esm/layouts/Split/index.js";
 
-import SerialConsole from './serialConsole.jsx';
-import Vnc from './vnc.jsx';
-import DesktopConsole from './desktopConsole.jsx';
-import {
-    domainCanConsole,
-    domainDesktopConsole,
-    domainSerialConsoleCommand
-} from '../../../libvirtApi/domain.js';
+import { ConsoleState } from './common';
+import { SerialActive, SerialInactive, SerialMissing, SerialPending } from './serial';
+import { VncActive, VncActiveActions, VncInactive, VncMissing, VncPending } from './vnc';
+import { SpiceActive, SpiceInactive } from './spice';
+
+import { domainSerialConsoleCommand } from '../../../libvirtApi/domain.js';
+import { vmId } from "../../../helpers.js";
 
 import './consoles.css';
 
 const _ = cockpit.gettext;
 
-const VmNotRunning = () => {
-    return (
-        <div id="vm-not-running-message">
-            {_("Please start the virtual machine to access its console.")}
-        </div>
-    );
-};
-
-class Consoles extends React.Component {
-    constructor (props) {
-        super(props);
-
-        this.state = {
-            serial: props.vm.displays && props.vm.displays.filter(display => display.type == 'pty'),
-        };
-
-        this.getDefaultConsole = this.getDefaultConsole.bind(this);
-        this.onDesktopConsoleDownload = this.onDesktopConsoleDownload.bind(this);
+class SerialStates extends StateObject {
+    constructor() {
+        super();
+        this.states = { };
     }
 
-    static getDerivedStateFromProps(nextProps, prevState) {
-        const oldSerial = prevState.serial;
-        const newSerial = nextProps.vm.displays && nextProps.vm.displays.filter(display => display.type == 'pty');
-
-        if (newSerial.length !== oldSerial.length || oldSerial.some((pty, index) => pty.alias !== newSerial[index].alias))
-            return { serial: newSerial };
+    get(key) {
+        if (!(key in this.states)) {
+            const state = new ConsoleState();
+            this.follow(state);
+            this.states[key] = state;
```

Comment:
These 5 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2008-57844738-20250716-080045-fedora-42-devel/Coverage/lcov/github-pr.diff.gcov.html)